src/box/module_cache.c

```diff
@@ -296,8 +296,8 @@ module_new(const char *package, size_t package_len,
 
 	memcpy(m->package, package, package_len);
 	m->package[package_len] = 0;
-
-	const char *tmpdir = getenv("TMPDIR");
+	bool exceeded;
+	const char *tmpdir = getenv_limited("TMPDIR", PATH_MAX, &exceeded);
```

Comment:
I don't like that using `getenv_limited` implies defining `exceeded` on stack, even if it's not required. This complicates usage of this function. And even in places, where we do check `exceeded`, the code after the patch looks just as complicated as before. Maybe, it'd be better to remove this argument and simply print a warning to the log in case the environment variable is too long? (Something like "Ignoring the value of environment variable XXX because it's too big (>= XXX bytes)".) Or would it break any expectations? I'd also call this function `getenv_safe` to emphasize that it's goal is security/safety.

Reply:
Ok, let it be `getenv_safe`.

Speaking of the `exceeded` argument, it's needed because some places (mostly the ones that were safe even without my patch) behave differently in case an env value is absent and in case it is too long.

I don't like that the `exceeded` is used everywhere too. I think I can make it optional.

Reply:
BTW, starting with C11 standard there's `getenv_s` library function with amost identical semantics. Can we use that instead?
https://en.cppreference.com/w/c/program/getenv

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/console.c b/src/box/lua/console.c
index 2987fb9bd..c45d8f0b7 100644
--- a/src/box/lua/console.c
+++ b/src/box/lua/console.c
@@ -391,9 +391,8 @@ lbox_console_show_prompt(struct lua_State *L)
 static bool
 console_hide_show_prompt_is_enabled(void)
 {
-	bool exceeded;
-	const char *envvar = getenv_limited("TT_CONSOLE_HIDE_SHOW_PROMPT", 10,
-					    &exceeded);
+	const char *envvar = getenv_safe("TT_CONSOLE_HIDE_SHOW_PROMPT", 10,
+					 NULL);
 
 	/* Enabled by default. */
 	if (envvar == NULL || *envvar == '\0')
diff --git a/src/box/module_cache.c b/src/box/module_cache.c
index 886313adc..1e4dbedd3 100644
--- a/src/box/module_cache.c
+++ b/src/box/module_cache.c
@@ -296,8 +296,7 @@ module_new(const char *package, size_t package_len,
 
 	memcpy(m->package, package, package_len);
 	m->package[package_len] = 0;
-	bool exceeded;
-	const char *tmpdir = getenv_limited("TMPDIR", PATH_MAX, &exceeded);
+	const char *tmpdir = getenv_safe("TMPDIR", PATH_MAX, NULL);
 	if (tmpdir == NULL)
 		tmpdir = "/tmp";
 
diff --git a/src/box/sql/os_unix.c b/src/box/sql/os_unix.c
index 79e38b06e..1d3fcc44e 100644
--- a/src/box/sql/os_unix.c
+++ b/src/box/sql/os_unix.c
@@ -1436,12 +1436,11 @@ unixTempFileDir(void)
 	unsigned int i = 0;
 	struct stat buf;
 	const char *zDir = sql_temp_directory;
-	bool exceeded;
 
 	if (!azDirs[0])
-		azDirs[0] = getenv_limited("SQL_TMPDIR", PATH_MAX, &exceeded);
+		azDirs[0] = getenv_safe("SQL_TMPDIR", PATH_MAX, NULL);
 	if (!azDirs[1])
-		azDirs[1] = getenv_limited("TMPDIR", PATH_MAX, &exceeded);
+		azDirs[1] = getenv_safe("TMPDIR", PATH_MAX, NULL);
 	while (1) {
 		if (zDir != 0 && stat(zDir, &buf) == 0 &&
 		    S_ISDIR(buf.st_mode) && access(zDir, 03) == 0)
diff --git a/src/lib/core/coio_file.c b/src/lib/core/coio_file.c
index efb2575f7..cdefe3b41 100644
--- a/src/lib/core/coio_file.c
+++ b/src/lib/core/coio_file.c
@@ -507,7 +507,7 @@ coio_tempdir(char *path, size_t path_len)
 	INIT_COEIO_FILE(eio);
 
 	bool exceeded;
-	const char *tmpdir = getenv_limited("TMPDIR", path_len, &exceeded);
+	const char *tmpdir = getenv_safe("TMPDIR", path_len, &exceeded);
 	if (exceeded) {
 		errno = ENOMEM;
 		return -1;
diff --git a/src/lib/core/errinj.c b/src/lib/core/errinj.c
index 348ce4ba2..9a5bc33f8 100644
--- a/src/lib/core/errinj.c
+++ b/src/lib/core/errinj.c
@@ -74,8 +74,7 @@ void errinj_set_with_environment_vars(void) {
 		struct errinj *inj = &errinjs[i];
 		bool exceeded;
 		/* 64 bytes ought to be enough for anybody. */
-		const char *env_value = getenv_limited(inj->name, 64,
-						       &exceeded);
+		const char *env_value = getenv_safe(inj->name, 64, &exceeded);
 		if (exceeded) {
 			panic("Too long value for %s passed via environment",
 			      inj->name);
diff --git a/src/lib/core/util.c b/src/lib/core/util.c
index 8930f90c2..44af20063 100644
--- a/src/lib/core/util.c
+++ b/src/lib/core/util.c
@@ -486,13 +486,14 @@ thread_sleep(double sec)
 }
 
 char *
-getenv_limited(const char *name, size_t buf_size, bool *exceeded)
+getenv_safe(const char *name, size_t buf_size, bool *exceeded)
 {
 	char *var = getenv(name);
 	if (var != NULL && strnlen(var, buf_size) >= buf_size) {
-		*exceeded = true;
+		if (exceeded != NULL)
+			*exceeded = true;
 		var = NULL;
-	} else {
+	} else if (exceeded != NULL) {
 		*exceeded = false;
 	}
 	return var;
diff --git a/src/lua/init.c b/src/lua/init.c
index 195490641..e4b99d154 100644
--- a/src/lua/init.c
+++ b/src/lua/init.c
@@ -572,8 +572,7 @@ fiber_backtracer(void *(*frame_writer)(int frame_no, void *addr))
 static void
 tarantool_lua_pushpath_env(struct lua_State *L, const char *envname)
 {
-	bool exceeded;
-	const char *path = getenv_limited(envname, PATH_MAX, &exceeded);
+	const char *path = getenv_safe(envname, PATH_MAX, NULL);
 	if (path != NULL) {
 		const char *def = lua_tostring(L, -1);
 		path = luaL_gsub(L, path, ";;", ";\1;");
@@ -590,8 +589,7 @@ tarantool_lua_pushpath_env(struct lua_State *L, const char *envname)
 static void
 tarantool_lua_setpaths(struct lua_State *L)
 {
-	bool exceeded;
-	const char *home = getenv_limited("HOME", PATH_MAX, &exceeded);
+	const char *home = getenv_safe("HOME", PATH_MAX, NULL);
 	lua_getglobal(L, "package");
 	int top = lua_gettop(L);
 
diff --git a/src/systemd.c b/src/systemd.c
index e0008810d..279e920d4 100644
--- a/src/systemd.c
+++ b/src/systemd.c
@@ -57,8 +57,8 @@ int systemd_init(void) {
 		.sun_family = AF_UNIX,
 		.sun_path = { '\0' }
 	};
-	sd_unix_path = getenv_limited("NOTIFY_SOCKET", sizeof(sa.sun_path),
-				      &exceeded);
+	sd_unix_path = getenv_safe("NOTIFY_SOCKET", sizeof(sa.sun_path),
+				   &exceeded);
 	if (exceeded) {
 		say_error("systemd: NOTIFY_SOCKET is longer than "
 			  "MAX_UNIX_PATH");
diff --git a/src/trivia/util.h b/src/trivia/util.h
index f258e8f57..c2683aafd 100644
--- a/src/trivia/util.h
+++ b/src/trivia/util.h
@@ -648,7 +648,7 @@ thread_sleep(double sec);
  * buffer of the given size.
  */
 char *
-getenv_limited(const char *name, size_t buf_size, bool *exceeded);
+getenv_safe(const char *name, size_t buf_size, bool *exceeded);
 
 #if !defined(__cplusplus) && !defined(static_assert)
 # define static_assert _Static_assert
```
</details>

Reply:
I like `getenv_s`. It'd be nice if we could use it if it exists and define it if it doesn't. However, I'm not sure it's compatible with our code - we often use environment variables without copying their values to a buffer.

Re the optional `exceeded` argument: I don't like that we silently ignore long environment variable values in quite a few places. This may be unexpected. It'd be nice to at least say something to the log.

> Speaking of the `exceeded` argument, it's needed because some places (mostly the ones that were safe even without my patch) behave differently in case an env value is absent and in case it is too long.

Maybe, we could leave those places as is, since they are safe anyway?

Reply:
Re the warning: adding `getenv_safe` to the message is redundant, because we add file/line to all warning messages anyway, and the name of the function gives no useful info to the user (the message is already self-sufficient).

Re using `getenv_safe` everywhere: it sounds like a good idea, but now it prints a warning message to the log, which is redundant in case of `systemd_init`, because it raises an error in this case. And because of `systemd_init`, we have to complicate `getenv_safe`.

Reply:
> Re the warning: adding getenv_safe to the message is redundant, because we add file/line to all warning messages anyway, and the name of the function gives no useful info to the user (the message is already self-sufficient).

Ok:
<details><summary>diff</summary>

```diff
diff --git a/src/lib/core/util.c b/src/lib/core/util.c
index d4b07e5ec..c00f6be11 100644
--- a/src/lib/core/util.c
+++ b/src/lib/core/util.c
@@ -492,7 +492,7 @@ getenv_safe(const char *name, size_t buf_size, bool *exceeded)
        if (var != NULL && strnlen(var, buf_size) >= buf_size) {
                if (exceeded != NULL)
                        *exceeded = true;
-               say_warn("getenv_safe: ignoring environment variable %s "
+               say_warn("Ignoring environment variable %s "
                         "because its value is longer, than expected. "
                         "Expected %zu bytes.", name, buf_size);
                var = NULL;
```
</details>

> Re using getenv_safe everywhere: it sounds like a good idea, but now it prints a warning message to the log, which is redundant in case of systemd_init, because it raises an error in this case. And because of systemd_init, we have to complicate getenv_safe.

IMO it'd be better for `getenv_safe` to print the same warnings regardless of `exceeded` flag being passed or not.
I'd probably just go with an extraneous warning.